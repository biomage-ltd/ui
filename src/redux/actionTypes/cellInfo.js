const CELL_INFO = 'cellInfo';

/**
 * Update cell information on currently highlighted cell.
 */
const CELL_INFO_UPDATE = `${CELL_INFO}/update`;

/**
 * Set focus on a particular metadata/cell set/gene.
 */
const CELL_INFO_FOCUS = `${CELL_INFO}/focus`;

/**
 * Set focus on a particular cell set or gene.
 */

const CELL_INFO_UNFOCUS = `${CELL_INFO}/unfocus`;

export {
  CELL_INFO_UPDATE, CELL_INFO_FOCUS, CELL_INFO_UNFOCUS,
};
