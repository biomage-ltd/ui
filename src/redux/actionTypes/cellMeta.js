const CELL_META = 'cellMeta';

/**
 * Load plot data
 */
const CELL_META_LOADING = `${CELL_META}/loading`;

/**
 * Plot data loaded
 */
const CELL_META_LOADED = `${CELL_META}/loaded`;

/**
 * Plot data error
 */
const CELL_META_ERROR = `${CELL_META}/error`;

export { CELL_META_LOADING, CELL_META_LOADED, CELL_META_ERROR };
