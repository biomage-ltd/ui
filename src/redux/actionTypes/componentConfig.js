/**
 * Sets embedding categorical config.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;

export { LOAD_CONFIG, UPDATE_CONFIG, SAVE_CONFIG };
