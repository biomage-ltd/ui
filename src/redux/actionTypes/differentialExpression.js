const DIFF_EXPR = 'diffExpr';

/**
 * Turns on the loading condition for differential expression.
 * Components that have the particular properties requested in their dependencies
 * should set a loading state at this event.
 */
const DIFF_EXPR_LOADING = `${DIFF_EXPR}/loading`;

/**
 * Sets the state of the differential expression data to be successfully loaded.
 * This is meant to be used for paginated data, where the data is
 * lazy-loaded by an API call or equivalent.
 */
const DIFF_EXPR_LOADED = `${DIFF_EXPR}/loaded`;

const DIFF_EXPR_ERROR = `${DIFF_EXPR}/error`;

/**
 * Sets the state for diffential expression settings
 *
 */
const DIFF_EXPR_COMPARISON_TYPE_SET = `${DIFF_EXPR}/comparisonType`;
const DIFF_EXPR_COMPARISON_GROUP_SET = `${DIFF_EXPR}/comparisonGroup`;

export {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
  DIFF_EXPR_COMPARISON_TYPE_SET, DIFF_EXPR_COMPARISON_GROUP_SET,
};
