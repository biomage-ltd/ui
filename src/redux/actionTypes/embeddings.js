const EMBEDDINGS = 'embeddings';

/**
 * Turns on the loading condition for the embedding.
 */
const EMBEDDINGS_LOADING = `${EMBEDDINGS}/loading`;

/**
 * Sets the state of the embedding to be successfully loaded, with the appropriate embedding data.
 */
const EMBEDDINGS_LOADED = `${EMBEDDINGS}/loaded`;

/**
 * Sets an error condition for a specific embedding.
 */
const EMBEDDINGS_ERROR = `${EMBEDDINGS}/error`;

export {
  EMBEDDINGS_LOADING, EMBEDDINGS_LOADED, EMBEDDINGS_ERROR,
};
