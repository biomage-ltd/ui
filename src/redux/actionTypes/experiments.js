const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATE = `${EXPERIMENTS}/create`;

/**
 * Update experiments
 */
const EXPERIMENTS_UPDATE = `${EXPERIMENTS}/update`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

/**
 * Experiment saved
 */
const EXPERIMENTS_SAVED = `${EXPERIMENTS}/saved`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

export {
  EXPERIMENTS_CREATE,
  EXPERIMENTS_UPDATE,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SAVED,
};
