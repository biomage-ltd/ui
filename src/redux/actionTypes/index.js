const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC,
  SET_FOCUSED_GENE,
};
