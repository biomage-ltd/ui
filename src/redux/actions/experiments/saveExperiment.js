/* eslint-disable no-param-reassign */
import fetchAPI from '../../../utils/fetchAPI';
import pushNotificationMessage from '../../../utils/pushNotificationMessage';
import {
  EXPERIMENTS_ERROR,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SAVED,
} from '../../actionTypes/experiments';
import errorTypes from './errorTypes';

const saveExperiment = (
  experimentId,
  newExperiment,
  alreadyExists = true,
) => async (dispatch, getState) => {
  const payload = newExperiment || getState().experiments[experimentId];

  dispatch({
    type: EXPERIMENTS_SAVING,
  });

  try {
    const response = await fetchAPI(
      `/v1/experiments/${experimentId}`,
      {
        method: alreadyExists ? 'PUT' : 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
      },
    );

    if (!response.ok) {
      throw new Error('HTTP status code was not 200.');
    }

    dispatch({
      type: EXPERIMENTS_SAVED,
    });
  } catch (e) {
    dispatch({
      type: EXPERIMENTS_ERROR,
      payload: {
        error: errorTypes.SAVE_EXPERIMENT,
      },
    });

    pushNotificationMessage(
      'error',
      'We couldn\'t connect to the server to save your current experiment, retrying...',
    );
  }
};

export default saveExperiment;
