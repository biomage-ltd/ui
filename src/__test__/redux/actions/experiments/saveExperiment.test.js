import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import initialExperimenttState, { experimentTemplate } from '../../../../redux/reducers/experiments/initialState';
import saveExperiment from '../../../../redux/actions/experiments/saveExperiment';
import {
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SAVED,
  EXPERIMENTS_ERROR,
} from '../../../../redux/actionTypes/experiments';
import { NOTIFICATIONS_PUSH_MESSAGE } from '../../../../redux/actionTypes/notifications';

jest.mock('localforage');

enableFetchMocks();
const mockStore = configureStore([thunk]);

describe('saveExperiment action', () => {
  const mockExperiment = {
    ...experimentTemplate,
    id: 'experiment-1',
    name: 'experiment-1',
    description: 'Project',
  };

  const initialState = {
    experiments: {
      ...initialExperimenttState,
      ids: [mockExperiment.id],
      [mockExperiment.id]: mockExperiment,
    },
  };

  beforeEach(() => {
    const response = new Response(JSON.stringify({ one: 'one' }));

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValueOnce(response);
  });

  it('Dispatches fetch correctly.', async () => {
    const store = mockStore(initialState);
    await store.dispatch(saveExperiment(mockExperiment.id));

    expect(fetchMock).toHaveBeenCalledWith(
      `http://localhost:3000/v1/experiments/${mockExperiment.id}`,
      {
        body: JSON.stringify(mockExperiment),
        headers: {
          'Content-Type': 'application/json',
        },
        method: 'PUT',
      },
    );
  });

  it('Dispatches correct actions when action is successful', async () => {
    const store = mockStore(initialState);

    await store.dispatch(saveExperiment(mockExperiment.id));

    const actions = store.getActions();

    // First action sets up saving status
    expect(actions[0].type).toBe(EXPERIMENTS_SAVING);

    // Second state saves error
    expect(actions[1].type).toBe(EXPERIMENTS_SAVED);

    expect(actions).toMatchSnapshot();
  });

  it('Dispatches a notification when fetch fails.', async () => {
    fetchMock.resetMocks();
    fetchMock.mockReject(new Error('some weird error that happened'));

    const store = mockStore(initialState);
    await store.dispatch(saveExperiment(mockExperiment.id));

    const actions = store.getActions();

    // First action sets up saving status
    expect(actions[0].type).toBe(EXPERIMENTS_SAVING);

    // Second state saves error
    expect(actions[1].type).toBe(EXPERIMENTS_ERROR);

    // Thirdd state emits notification
    expect(actions[2].type).toBe(NOTIFICATIONS_PUSH_MESSAGE);

    expect(actions).toMatchSnapshot();
  });
});
