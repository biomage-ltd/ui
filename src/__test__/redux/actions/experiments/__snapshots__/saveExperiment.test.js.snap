// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saveExperiment action Dispatches a notification when fetch fails. 1`] = `
Array [
  Object {
    "type": "experiments/saving",
  },
  Object {
    "payload": Object {
      "error": "Error saving experiment",
    },
    "type": "experiments/error",
  },
]
`;

exports[`saveExperiment action Dispatches correct actions when action is successful 1`] = `
Array [
  Object {
    "type": "experiments/saving",
  },
  Object {
    "type": "experiments/saved",
  },
]
`;
